crates/oxc_linter/src/plugin.rs

```diff
@@ -0,0 +1,233 @@
+use std::{collections::BTreeMap, env, fs, rc::Rc, sync::Arc};
+
+use oxc_diagnostics::miette::{miette, LabeledSpan};
+use oxc_query::Adapter;
+use oxc_semantic::Semantic;
+use serde::Deserialize;
+use trustfall::{execute_query, Schema, TransparentValue, TryIntoStruct};
+
+use crate::context::LintContext;
+
+enum SpanInfo {
+    SingleSpanInfo(SingleSpanInfo),
+    MultipleSpanInfo(MultipleSpanInfo),
+}
+
+#[derive(Debug, Deserialize)]
+struct SingleSpanInfo {
+    span_start: u64,
+    span_end: u64,
+}
+
+#[derive(Debug, Deserialize)]
+struct MultipleSpanInfo {
+    span_start: Box<[u64]>,
+    span_end: Box<[u64]>,
+}
+
+#[derive(Deserialize, Clone)]
+pub struct InputQuery {
+    pub name: String,
+    pub query: String,
+    pub args: BTreeMap<Arc<str>, TransparentValue>,
+    pub reason: String,
+    #[serde(default)]
+    pub tests: QueryTests,
+}
+
+#[derive(Deserialize, Default, Clone)]
+pub struct QueryTests {
+    pub pass: Vec<SingleTest>,
+    pub fail: Vec<SingleTest>,
+}
+
+#[derive(Deserialize, Clone)]
+pub struct SingleTest {
+    pub relative_path: Vec<String>,
+    pub code: String,
+}
+
+pub struct LinterPlugin {
+    pub(super) rules: Vec<InputQuery>,
+    schema: &'static Schema,
+}
+
+impl LinterPlugin {
+    pub fn new(schema: &'static Schema) -> Self {
+        let queries_path = env::var("OXC_PLUGIN").unwrap();
+        let rules = fs::read_dir(queries_path)
```

Comment:
this is brittle code that enforces each rule has its own folder with subrules, really we just need to recurse downard from the env var looking for files that look like what we need